### PR TITLE
Changes to pass most tests on Oracle

### PR DESCRIPTION
**Pull request by srinicodebytes**
- Add separate insert statements.
  - Skip tests that use unsupported data types.
  - 116 tests out of 117 pass in DatabaseAdaptorTest
  - 35 test out of 43 pass in ResponseGeneratorTest

### CODE REVIEW

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 1]
```diff
@@ -113,7 +115,7 @@ private String nowPlus(int minutes) {
       case MYSQL:
         return "date_add(current_timestamp(), interval " + minutes + " minute)";
       case ORACLE:
-        // TODO(sv): minute_add???
+        return "(select sysdate + interval '" + minutes + "' minute from dual)";
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: This is just an expression, so why do you need to select it from dual?

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 2]
```diff
@@ -1371,8 +1383,10 @@ public void testInitVerifyColumnNames_sqlException() throws Exception {
   @Test
   public void testGetDocIds() throws Exception {
     executeUpdate("create table data(id integer, other varchar(20))");
-    executeUpdate("insert into data(id, other) values(1, 'hello world'),"
-        + "(2, 'hello world'), (3, 'hello world'), (4, 'hello world')");
+    executeUpdate("insert into data(id, other) values(1, 'hello world')");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java:1400]: executeUpdate is a varargs method. Not sure whether that's more readable or not for these repetitive calls.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 3]
```diff
@@ -2159,7 +2193,11 @@ public void testMetadataColumns_date() throws Exception {
 
     Metadata metadata = new Metadata();
     metadata.add("col1", "1001");
-    metadata.add("col2", "2004-10-06");
+    if (is(ORACLE)) {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: metadata.add("col2", is(ORACLE) ? "2004-10-06 00:00:00.0" : "2004-10-06");

File: test/com/google/enterprise/adaptor/database/ResponseGeneratorTest.java [hunk 4]
```diff
@@ -289,6 +292,7 @@ public void testContentColumn_incorrectColumnName() throws Exception {
 
   @Test
   public void testContentColumn_varbinary() throws Exception {
+    assumeFalse("Oracle does not support varbinary", is(ORACLE));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/ResponseGeneratorTest.java]: Try swizzling varbinary to raw in JdbcFixture.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 5]
```diff
@@ -419,7 +421,12 @@ public void testAclSqlResultSetHasNonOverlappingTwoRecord()
         + GsaSpecialColumns.GSA_DENY_GROUPS + ","
         + GsaSpecialColumns.GSA_DENY_USERS + ") values "
         + "('pgroup1, pgroup2', 'puser1, puser2', "
-        + "'dgroup1, dgroup2', 'duser1, duser2'), "
+        + "'dgroup1, dgroup2', 'duser1, duser2')");
+    executeUpdate("insert into acl("
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java:439]: For posterity, these would have made a good separate CL. It's hard to scan for other changes here.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 6]
```diff
@@ -2396,16 +2436,29 @@ public void testMetadataColumns_blobNull() throws Exception {
   }
 
   @Test
-  public void testMetadataColumns_array() throws Exception {
-    assumeFalse("MySQL does not support arrays", is(MYSQL));
+  public void testMetadataColumns_array_Oracle() throws Exception {
+    assumeTrue(is(ORACLE));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: I think it might be more readable and true to the spirit of these tests to just use an if in the one test method:

if (is(ORACLE)) { ... } else if (is(H2)) { ... } else { assumeTrue(false); }

I don't know if there's a better way to say assumeTrue(false).

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 1 above)
Comment (BrettMichaelJohnson) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Wouldn't it be: return "(CURRENT_TIMESTAMP + interval '" + minutes + "' minute)";

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 3 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 1 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 5 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java:439]: Ok.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 2 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java:1400]: I think, this is more readable.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 6 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done.

File: test/com/google/enterprise/adaptor/database/JdbcFixture.java [hunk 7]
```diff
@@ -179,6 +179,18 @@ public static void executeUpdate(String... sqls) throws SQLException {
     try (Connection connection = getConnection();
          Statement stmt = connection.createStatement()) {
       for (String sql : sqls) {
+        switch (DATABASE) {
+          case ORACLE:
+            if (sql.startsWith("create table")) {
+              sql = sql.replaceAll("varbinary", "raw(1024)");
+            }
+            if (sql.startsWith("insert")) {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/JdbcFixture.java]: Why is this needed for Oracle?

File: test/com/google/enterprise/adaptor/database/ResponseGeneratorTest.java [hunk 8]
```diff
@@ -557,7 +561,7 @@ public void testUrlAndMetadataLister_ignoresColumnName() throws Exception {
   @Test
   public void testUrlAndMetadataLister_nullUrl() throws Exception {
     executeUpdate("create table data(url varchar(200))");
-    executeUpdate("insert into data() values ()");
+    executeUpdate("insert into data(url) values (NULL)");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/ResponseGeneratorTest.java]: s/NULL/null/

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 9]
```diff
@@ -2394,12 +2431,21 @@ public void testMetadataColumns_blobNull() throws Exception {
   @Test
   public void testMetadataColumns_array() throws Exception {
     assumeTrue("ARRAY type not supported", is(H2));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: This needs patched up as I described before.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 10]
```diff
@@ -2426,8 +2472,15 @@ public void testMetadataColumns_array() throws Exception {
   @Test
   public void testMetadataColumns_arrayNull() throws Exception {
     assumeTrue("ARRAY type not supported", is(H2));
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Same here.

File: test/com/google/enterprise/adaptor/database/ResponseGeneratorTest.java (hunk 8 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/ResponseGeneratorTest.java]: Done.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 9 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 10 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 11]
```diff
@@ -2471,15 +2472,16 @@ public void testMetadataColumns_array() throws Exception {
 
   @Test
   public void testMetadataColumns_arrayNull() throws Exception {
-    assumeTrue("ARRAY type not supported", is(H2));
-    if (is(ORACLE)) {
+    if(is(H2)) {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Add space after "if".

File: test/com/google/enterprise/adaptor/database/JdbcFixture.java [hunk 12]
```diff
@@ -211,6 +211,11 @@ public static void executeUpdate(String... sqls) throws SQLException {
             }
             break;
         }
+        if (sql.startsWith("insert")) {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/JdbcFixture.java]: Missed this. Please restore to MySQL only.

**Comment by wiarlawd** [comment]
While I'm thinking of it, s/tets/tests/ in the description.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java [hunk 13]
```diff
@@ -2410,13 +2442,23 @@ public void testMetadataColumns_blobNull() throws Exception {
...
-    executeUpdate("create table data(id int, content array)");
-    String sql = "insert into data(id, content) values (1, ?)";
-    PreparedStatement ps = prepareStatement(sql);
-    ps.setObject(1, content);
-    assertEquals(1, ps.executeUpdate());
+    if (is(H2)) {
+      String[] content = {"hello", "world"};
+      executeUpdate("create table data(id int, content array)");
+      String sql = "insert into data(id, content) values (1, ?)";
+      try (PreparedStatement ps = getConnection().prepareStatement(sql)) {
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Brett deleted this try statement on master, and the merge missed that. Restore this block to indentation-only changes.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 11 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done.

**Review by wiarlawd — APPROVED** [review]
LGTM++ with a merge fix.

File: test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java (hunk 13 above)
Comment (srinicodebytes) [reply test/com/google/enterprise/adaptor/database/DatabaseAdaptorTest.java]: Done.
